internal/communicator/ssh/http_proxy.go

```diff
@@ -0,0 +1,168 @@
+package ssh
+
+import (
+	"bufio"
+	"encoding/base64"
+	"fmt"
+	"net"
+	"net/http"
+	"net/url"
+	"time"
+
+	"golang.org/x/net/proxy"
+)
+
+// Dialer implements for SSH over HTTP Proxy.
+type Dialer struct {
+	proxy proxyInfo
+	// forwarding Dialer
+	forward proxy.Dialer
+}
+
+type proxyInfo struct {
+	// HTTP Proxy host or host:port
+	host string
+	// HTTP Proxy scheme
+	scheme string
+	// User name if http proxy needs authentication
+	username string
+	// User password if http proxy needs authentication
+	password string
+	// Whether the HTTP Proxy requires authentication
+	auth bool
+}
+
+func newProxyInfo(host, scheme, username, password string) *proxyInfo {
+	p := &proxyInfo{
+		host:   host,
+		scheme: scheme,
+	}
+
+	if username != "" && password != "" {
+		p.auth = true
+		p.username = username
+		p.password = password
+	}
+
+	if p.scheme == "" {
+		p.scheme = "http"
+	}
+
+	return p
+}
+
+func (p *proxyInfo) url() (*url.URL, error) {
+	base := p.scheme + "://"
+
+	if p.auth {
+		base = base + p.username + ":" + p.password + "@"
+	}
+
+	return url.Parse(base + p.host)
+}
+
+func (p *Dialer) Dial(network, addr string) (net.Conn, error) {
+	// Dial the proxy host
+	c, err := p.forward.Dial(network, p.proxy.host)
+
+	if err != nil {
+		return nil, err
+	}
+
+	err = c.SetDeadline(time.Now().Add(15 * time.Second))
+	if err != nil {
+		return nil, err
+	}
+
+	// Generate request URL to host accessed through the proxy
+	reqUrl, err := url.Parse("http://" + addr)
+	if err != nil {
+		c.Close()
+		return nil, err
+	}
+	reqUrl.Scheme = ""
+
+	// Create a request object using the CONNECT method to instruct the proxy server to tunnel a protocol other than HTTP.
+	req, err := http.NewRequest("CONNECT", reqUrl.String(), nil)
+	if err != nil {
+		c.Close()
+		return nil, err
+	}
+
+	// If http proxy requires authentication, configure settings for basic authentication.
+	if p.proxy.auth {
+		req.SetBasicAuth(p.proxy.username, p.proxy.password)
+		req.Header.Add("Proxy-Authorization", "Basic "+base64.StdEncoding.EncodeToString([]byte(p.proxy.username+":"+p.proxy.password)))
```

Comment:
Instead of building this value ourselves, could we reuse the authorization header value?
```suggestion
		req.Header.Add("Proxy-Authorization", req.Header.Get("Authorization"))
```

Reply:
Fixed